File: 542 01 Matrix/542 01 Matrix.go
func updateMatrix(mat [][]int) [][]int {
    q := list.New()
    m, n := len(mat), len(mat[0])
    for i := 0; i < m; i++ {
        for j := 0; j < n; j++ {
            if mat[i][j] == 0 {
                q.PushBack([2]int{i, j})
            } else {
                mat[i][j] = -1
            }
        }
    }
    step := 0
    dirs := [4][2]int{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}
    for q.Len() > 0 {
        k := q.Len()
        for k > 0 {
            item := q.Remove(q.Front()).([2]int)
            for _, d := range dirs {
                x, y := item[0] + d[0], item[1] + d[1]
                if x >= 0 && x < m && y >= 0 && y < n && mat[x][y] == -1 {
                    mat[x][y] = step+1
                    q.PushBack([2]int{x, y})
                }
            }
            k--
        }
        step++
    }
    return mat
}

File: 542 01 Matrix/542 01 Matrix.py
"""
Given an m x n binary matrix mat, return the distance of the nearest 0 for each cell.
The distance between two adjacent cells is 1.

Example 1:

Input: mat = [[0,0,0],[0,1,0],[0,0,0]]
Output: [[0,0,0],[0,1,0],[0,0,0]]

Example 2:

Input: mat = [[0,0,0],[0,1,0],[1,1,1]]
Output: [[0,0,0],[0,1,0],[1,2,1]]
"""
class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        if mat[0][0] == 1:
            mat[0][0] = float('inf')

        for i in range(1, len(mat[0])):
            if mat[0][i] == 1:
                mat[0][i] = mat[0][i-1] + 1

        for i in range(1, len(mat)):
            if mat[i][0] == 1:
                mat[i][0] = mat[i-1][0] + 1

        for i in range(1, len(mat)):
            for j in range(1, len(mat[0])):
                if mat[i][j] == 1:
                    mat[i][j] = min(mat[i-1][j], mat[i][j-1]) + 1

        for i in range(len(mat[0])-2, -1, -1):
            if mat[-1][i] != 0:
                mat[-1][i] = min(mat[-1][i+1] + 1, mat[-1][i])

        for i in range(len(mat)-2, -1, -1):
            if mat[i][-1] != 0:
                mat[i][-1] = min(mat[i+1][-1] + 1, mat[i][-1])

        for i in range(len(mat)-2, -1, -1):
            for j in range(len(mat[0])-2, -1, -1):
                if mat[i][j] != 0:
                    mat[i][j] = min(min(mat[i+1][j], mat[i][j+1])+1, mat[i][j])
        return mat
